### DISCUSSION TITLE
How to shorten the path of JustJ for Eclipse RCP applications

### DISCUSSION THREAD

**Discussion by vogella** [eclipse-justj/justj#16] [Q&A]
Is it possible to shorten the path of JustJ during the Tycho build? 

On Windows I have a client who runs into the folder length restriction with the long JustJ folder name. Is there a standard way to instruct Tycho to shorten the folder name and use the correct reference in the eclipse.ini file?

**Comment by merks** [comment]
This sounds like a Tycho question. 

Note that I build the Eclipse Installer on Windows without path length limitation, though some of the paths look **really** long:

> D:/Users/merks/oomph-1.30/git/oomph/products/org.eclipse.oomph.setup.installer.product/target/products/org.eclipse.oomph.setup.installer.product.with-jre.restricted/win32/win32/x86_64/plugins/org.eclipse.justj.openjdk.hotspot.jre.minimal.stripped.win32.x86_64_17.0.8.v20230801-1951/jre/org.eclipse.justj.properties

I don't recall enabling long paths but this does look longer than 256...

And I can make even longer paths:

> $touch "D:/Users/merks/oomph-1.30/git/oomph/products/org.eclipse.oomph.setup.installer.product/target/products/org.eclipse.oomph.setup.installer.product.with-jre.restricted/win32/win32/x86_64/plugins/org.eclipse.justj.openjdk.hotspot.jre.minimal.stripped.win32.x86_64_17.0.8.v20230801-1951/jre/org.eclipse.justj.properties.xadskljdsaklsdaffsafjkdsdhfdsadsfasfafjldasfdasfdsfadsfadsafdsfaadsfadsfjadsfadsfadsfadsfadsfjadsfadsfadsfkdsafjadsfadsfadsffasdfaadakdsfaajdsfadfadsfjsdfadsafdsfajdsaf"

**Comment by vogella** [comment]
On WIndows 10 the default path lenght is restricted: https://learn.microsoft.com/en-us/windows/win32/fileio/maximum-file-path-limitation?tabs=registry

The limit can be adjusted but requires admin rights. (The registry key Computer\HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\FileSystem\LongPathsEnabled (Type: REG_DWORD) ).

Are you using Windows11?

**Comment by merks** [comment]
I'm on Windows 10; haven't updated yet

I don't appear to have LongPathsEnabled enabled:

![image](https://github.com/eclipse-justj/justj/assets/208716/4c6d9121-5edc-41a0-9ae3-395318cfc641)

**Comment by vogella** [comment]
Not sure what is different for you, but if we try to start such a product with a reference to such a long JRE, we get "Failed to load the JNI shared library".

If we shorten the path and update the ini the application starts fine.

**Comment by merks** [comment]
It's hard to comment on "such a product".  Maybe you can compare it to this product:

https://download.eclipse.org/oomph/products/latest/eclipse-inst-jre-win64.zip

I'm also not sure exactly which path is being shortened.  Does the installer have an analogous such long path?

**Comment by vogella** [comment]
Thanks @merks for the example. This runs also for us even if we make the path much longer. We currently not sure what the difference is and are investigating now, if we find something we update the ticket.

**Comment by merks** [comment]
Best of luck!!  

Note too that all the EPP packages are also using a JustJ JRE, presumably with a similarly long path...

**Comment by pashor** [comment]
Thanks a lot for looking into this @merks!

I've downloaded the example ZIP file and re-tested it with a really long path name:
_"C:\test-JRE-long-path-name-EXCEPTION_test-JRE-with-exteremely-long-path-names-on-windows-operating-system-with-path-length-restrictions\test-JRE-with-exteremely-long-path-names-on-windows-operating-system-with-path-length-restrictions"_

In this case, I'm getting the below exception, when trying to start the `eclipse-inst.exe`:

> The Eclipse Installer executable launcher was unable to locate its companion shared library.


Interesting enough, if I set the registry key _Computer\HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Control\FileSystem\LongPathsEnabled_ to 1 and reboot, the problem still persists...

However, If I shorten the path-name to something like
_"C:\test-JRE-long-path-name\test-JRE-with-exteremely-long-path-names-on-windows-operating-system-with-path-length-restrictions"_
then the executable starts up fine.


my Execution Environment:
_Windows 10 Enterprise
Version 22H2 (OS Build 19045.3208)_

**Reply by merks** [reply]
I do recall from this issue being raised in the past that we are sometimes close to a limit.  

But if you look at the name, there is not a huge amount of savings to be had by shorting this part of the name `org.eclipse.justj.openjdk.hotspot.jre.minimal.stripped`.  Whatever we save here, a deep installation path can use up the limit...

Also note that these two paths are not so different in length:
```
D:\stuff\eclipse-inst-jre-win64(3).zip\plugins\org.eclipse.equinox.launcher.win32.win32.x86_64_1.2.700.v20221108-1024\eclipse_11801.dll
D:\stuff\eclipse-inst-jre-win64(3).zip\plugins\org.eclipse.justj.openjdk.hotspot.jre.minimal.stripped.win32.x86_64_17.0.8.v20230801-1951\jre\bin\java.exe
```

**Comment by pashor** [comment]
> Whatever we save here, a deep installation path can use up the limit...

Right.

In other words, we have to ask our Windows users to _not use_ a deep installation path (because even altering the "LongPathsEnabled" registry key seems to have no effect). This is acceptable.

The initial question was raised, because (for an end-user, not a developer) the error message seems to be "cryptic". At least not clear enough, i.e. it doesn't tell the user that the path is too long and it seems there is also no further logfile created that would contain more details. In such case, the end-user would just call the tool developer and complain that the executable doesn't start up.

But this, eventually, may be a thing to fix in the launcher code and not necessarily in the JustJ plug-in.
Is my assumption correct, @vogella and @merks?

**Comment by merks** [comment]
If there are failures very early, a log might be created in the configuration folder of the installation.  if the installation is read-only, there will be some type of surrogate in the ~.eclipse folder where the log will be in the configuration of that.   You could also try to run with -consoleLog.  I'm not sure what such an error looks like and how it's present to the user.  Likely that could be improved, but if this is very early in native start up code, that's probably not easy to improve...
